keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,652 @@
+because the workers often don't have any ways to terminate themselves due to launching daemon processes like ssh-server.
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successPolicy:
+  - target:
+      succeededIndexes: "0"
+      whenTargetAchieved: terminate
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./sample"]
+```
+
+#### Story 2
+
+As a simulation researcher/engineer for fluid dynamics/biochemistry.
+I want to mark the job as successful and continue running the rest of the pods if some indexes succeed
+because I set the minimum required value for sampling in the `.successPolicy.target.count` and
+perform the same simulation in all indexes.
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successPolicy:
+  - target:
+      count: 5
+      whenTargetAchieved: continue
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+        - name: job-container
+          image: job-image
+          command: ["./sample"]
+```
+
+### Story 3
+
+As a simulation researcher/engineer for fluid dynamics/biochemistry.
+I want to mark the job as successful and to perform the matched `whenTargetAchieved` of `successPolicy`
+when the job meets the one of following conditions:
+
+1. Once the leader index (index=0) is succeeded, I want to terminate running the rest of the pods.
+2. Once some worker indexes (index=1+) are succeeded, I want to continue running the rest of the pods for sampling.
+
+because succeeded leader index means that the whole simulation is succeeded,
+and succeeded some worker indexes means that the minimum required value for sampling is satisfied.
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successPolicy:
+  - target:
+      count: 1
+      index: 0
+      whenTargetAchieved: terminate
+  - target:
+      count: 5
+      index: "1-9"
+      whenTargetAchieved: continue
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+        - name: job-container
+          image: job-image
+          command: ["./sample"]
+```
+
+
+### Notes/Constraints/Caveats (Optional)
+
+#### No support JobSuccessPolicy for the NonIndexed Job
+
+As I described in [Non-Goals](#non-goals), we don't support the SuccessPolicy for the job with `NonIndexed` mode.
+
+#### Interchangeability between concepts of "completed" and "succeeded"
+In this KEP, we introduce the new `succeeded` concept, 
+which is interchangeable with the existing `completed` concept.
+
+It means that if the job meets the dedicated successPolicy, the `Completed` condition is added to the job.  
+
+#### Surprising whenTargetAchieved behavior if whenTargetAchieved is not specified somewhere  
```

Comment:
> In general, I'd advice to set a specific terminated or continue a default policy for a target.

Yes, we're planning to default `terminated` action, as mentioned in an API comment. This note built based on the following review comment.

> Here, it looks more like a room for error. Say we have a default whenTargetAchieved: terminate, and the user has two targets, but only sets the whenTargetAchieved value for one, as continue. Then, it would be defaulted in the other target. As a result, it would have two different values. IMO this would rarely be the intention.

https://github.com/kubernetes/enhancements/pull/4062#discussion_r1462132761

Reply:
@mimowo I considered your above comment, again. 
So, since once the job meets one of the successPolicy, the other successPolicy is ignored, and, I guess that this surprise wouldn't happen. 

WDYT?

Reply:
The defaulting sets it to a particular policy, so the API is consistent.

Users can always misconfigure the API, but that's not a problem of the design. I would just remove this section.

Reply:
SGTM
Thank you for the confirmation.

Reply:
Done.